peer.go

```diff
@@ -0,0 +1,1087 @@
+			stopCh:             make(chan struct{}),
+		},
+		controlCh:  controlCh,
+		progressCh: progressCh,
+
+		backoffTimer: time.NewTimer(time.Hour),
+		progress: peerProgress{
+			peerID: serverID,
+		},
+	}
+	p.backoffTimer.Stop()
+
+	pipeline, err := p.shared.trans.AppendEntriesPipeline(p.shared.peerAddr)
+	if err != nil {
+		if err != ErrPipelineReplicationNotSupported {
+			p.shared.logger.Printf("[ERR] raft: Failed to start pipeline replication to %s: %s",
+				p.shared.peerAddr, err)
+		}
+	} else {
+		p.shared.pipeline = pipeline
+	}
+
+	return p
+}
+
+func (p *peerState) checkInvariants() error {
+	if p.shared.peerID != p.progress.peerID {
+		return fmt.Errorf("Progress must have peerID")
+	}
+	if p.progress.lastContact.After(p.progress.lastReply) ||
+		p.progress.lastContact == p.progress.lastReply && !p.progress.lastContact.IsZero() {
+		return fmt.Errorf("lastContact must be after lastReply (or both 0)")
+	}
+	if p.nextIndex > p.control.lastIndex+1 {
+		return fmt.Errorf("nextIndex must not be more than 1 entry past the lastIndex")
+	}
+	if p.outstandingInstallSnapshotRPC && p.outstandingAppendEntriesRPCs > 0 {
+		return fmt.Errorf("must not send a snapshot and entries simultaneously")
+	}
+	return nil
+}
+
+// selectLoop is a long-running routine that sends RPCs to a single remote
+// server.
+func (p *peerState) selectLoop() {
+	for {
+		// At this point, we may or may not have more work that we could do right away.
+		// Check for new information but don't block.
+		p.drainNonblocking()
+		if p.control.role == Shutdown {
+			break
+		}
+		if err := p.checkInvariants(); err != nil {
+			panic("peer invariant violated: " + err.Error())
+		}
+		didSomething := p.issueWork()
+		// If we've got no more work to do, block until something changes.
+		if !didSomething {
+			p.blockingSelect()
+		}
+	}
+
+	p.shared.logger.Printf("[INFO] raft: Peer routine for %v exiting", p.shared.peerID)
+	close(p.shared.stopCh)
+	p.shared.pipeline.Close()
+}
+
+// drainNonblocking reads/writes the Peer channels as much as possible without
+// blocking.
+func (p *peerState) drainNonblocking() {
+	for {
+		// These cases should be exactly the same as those in blockingSelect, except
+		// with an additional default branch and without the heartbeat timer.
+		select {
+		case control := <-p.controlCh:
+			p.updateControl(control)
+		case p.progressChIfDirty() <- p.progress:
+			p.sendProgress = false
+		case request := <-p.shared.requestCh:
+			p.send(request)
+		case rpc := <-p.shared.replyCh:
+			p.processReply(rpc)
+		case <-p.shared.pipelineSendDoneCh:
+			p.outstandingPipelineSend = false
+		case <-p.backoffTimer.C:
+			p.failures = 0
+		default:
+			return
+		}
+	}
+}
+
+// blockingSelect reads/writes the Peer channels just once, blocking if needed.
+func (p *peerState) blockingSelect() {
+	// We need to send a heartbeat at lastHeartbeatSent + heartbeatInterval.
+	heartbeatTimer := time.After(p.shared.options.heartbeatInterval -
```

Comment:
Need an if statement around this so that Followers and Candidates don't fire all the time.